{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 01. Data Simulation\n",
    "\n",
    "This notebook simulates the amplification process. \n",
    "\n",
    "Steps: \n",
    "* Simulate data (data_simulator.py)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: data_simulator.py [-h] [--ado_poisson_rate ADO_POISSON_RATE]\n",
      "                         [--ado_type ADO_TYPE] [--amp_method AMP_METHOD]\n",
      "                         [--chr_id CHR_ID] [--genome_length GENOME_LENGTH]\n",
      "                         [--global_dir GLOBAL_DIR] [--gsnv_rate GSNV_RATE]\n",
      "                         [--is_flat IS_FLAT]\n",
      "                         [--mut_poisson_rate MUT_POISSON_RATE]\n",
      "                         [--num_cells NUM_CELLS] [--num_iter NUM_ITER]\n",
      "                         [--num_max_mid_points NUM_MAX_MID_POINTS]\n",
      "                         [--num_rep_amp_bias NUM_REP_AMP_BIAS] [--p_ado P_ADO]\n",
      "                         [--p_ae P_AE] [--phred_type PHRED_TYPE]\n",
      "                         [--read_length READ_LENGTH] [--seed_val SEED_VAL]\n",
      "\n",
      "Generate synthetic data (BAM) files.\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  --ado_poisson_rate ADO_POISSON_RATE\n",
      "                        Specify the ado poisson rate. Default: 0\n",
      "  --ado_type ADO_TYPE   Specify the allelic dropout type (0 for no ado, 1 for\n",
      "                        random Beta, 2 for even smaller Beta, 3 for fixed.\n",
      "                        Default: 3).\n",
      "  --amp_method AMP_METHOD\n",
      "                        Specify the amplification method. Default: mda\n",
      "  --chr_id CHR_ID       Specify the chromosome number. Default: 1\n",
      "  --genome_length GENOME_LENGTH\n",
      "                        Specify the length of genome. Default: 50000\n",
      "  --global_dir GLOBAL_DIR\n",
      "                        Specify the directory.\n",
      "  --gsnv_rate GSNV_RATE\n",
      "                        Specify the gSNV rate. Default: 0.005\n",
      "  --is_flat IS_FLAT     Specify the amplification bias type (True for flat,\n",
      "                        False for sine wave). Default: False\n",
      "  --mut_poisson_rate MUT_POISSON_RATE\n",
      "                        Specify the mutation poisson rate. Default: 3\n",
      "  --num_cells NUM_CELLS\n",
      "                        Specify the number of cells. Default: 10\n",
      "  --num_iter NUM_ITER   Specify the number of iteration. Default: 5000\n",
      "  --num_max_mid_points NUM_MAX_MID_POINTS\n",
      "                        Specify the number of division points of amplification\n",
      "                        bias. Default: 10\n",
      "  --num_rep_amp_bias NUM_REP_AMP_BIAS\n",
      "                        Specify the number of repetitions of amplification\n",
      "                        bias. Default: 3\n",
      "  --p_ado P_ADO         Specify the allelic dropout probability of a base.\n",
      "                        Default: 0.005\n",
      "  --p_ae P_AE           Specify the amplification error probability of a base.\n",
      "                        Default: 0.00001\n",
      "  --phred_type PHRED_TYPE\n",
      "                        Specify the phred score type (0 for no errors, 1 for\n",
      "                        all 42, 2 for truncated normal). Default: 2\n",
      "  --read_length READ_LENGTH\n",
      "                        Specify the read length. Default: 100\n",
      "  --seed_val SEED_VAL   Specify the seed. Default: 123\n"
     ]
    }
   ],
   "source": [
    "%run data_simulator.py --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The data directory:  ../data/2020_07_09_1/\n",
      "Directory is created:  ../data/2020_07_09_1/\n",
      "\n",
      "Part 0: Save arguments\n",
      "\n",
      "\n",
      "Total time:  0.0007691383361816406\n",
      "Part 0 ends...\n",
      "\n",
      "Data simulation starts...\n",
      "\n",
      "Part 1: Amplification bias\n",
      "\n",
      "\tMiddle points for repetition  0 :  [ 466  860 4426 5191 5390 5734 6265]\n",
      "\tMiddle points for repetition  1 :  [2064 2303 3392 7985 8499]\n",
      "\tMiddle points for repetition  2 :  [ 571 2144 3491 3971 9389]\n",
      "\n",
      "Total time:  0.2993500232696533\n",
      "Part 1 ends...\n",
      "\n",
      "Part 2: Reference, bulk and mutation generation\n",
      "\n",
      "\tnumBp:  10000\n",
      "\tgSNV count:  50\n",
      "\tmutation count:  250\n",
      "\tTotal gSNV count:  50 \tTotal mutation count:  250 \tMutations close to a gSNV:  126\n",
      "\tRunning:  samtools view ../data/2020_07_09_1/bulk_ref.sam | awk '{OFS=\"\\t\"; print \">\"$1\"\\n\"$10}' - > ../data/2020_07_09_1/bulk_ref.fasta\n",
      "\tRunning:  samtools view -h -b -S ../data/2020_07_09_1/bulk.sam > ../data/2020_07_09_1/bulk.bam\n",
      "\tRunning:  samtools index -b ../data/2020_07_09_1/bulk.bam\n",
      "\n",
      "Total time:  1.123128890991211\n",
      "Part 1 ends...\n",
      "\n",
      "Part 3: Tree generation and mutation partition\n",
      "\n",
      "\tparent nodes:  [0 0 0 1 1 2 2 5 5]\n",
      "\tleaf nodes:  [3, 4, 6, 7, 8]\n",
      "\n",
      "Total time:  0.001355886459350586\n",
      "Part 3 ends...\n",
      "\n",
      "Part 4: Cell simulation\n",
      "\n",
      "\n",
      "\tSimulating cell:  0\n",
      "\n",
      "\tNumber of Mutations:  58\n",
      "\n",
      "\tTotal time for cell's generateCellGenotype:  0.0008101463317871094\n",
      "\tp_ado:  0.3\n",
      "\tNumber of mask events:  63\n",
      "\t\tNumber of covered basepairs:  6327\n",
      "\t\tMask information:  {128: 1, 80: 1, 56: 1, 83: 2, 85: 2, 86: 1, 87: 1, 88: 2, 90: 1, 91: 1, 92: 5, 93: 2, 94: 2, 95: 2, 96: 1, 99: 2, 100: 3, 101: 2, 102: 4, 104: 7, 105: 1, 106: 2, 107: 1, 109: 3, 110: 1, 111: 3, 112: 1, 113: 1, 115: 1, 116: 1, 117: 2, 118: 1, 119: 1, 120: 1}\n",
      "\n",
      "\tTotal time for cell's maskGenome:  0.0023758411407470703\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.7407715320587158 0.10688328742980957 0.6338882446289062 0.6226072311401367\n",
      "\tNumber of rejected fragments due to amplification wave:  6470\n",
      "\n",
      "\tTotal number of fragments:  790\n",
      "\t\tNumber of fragments without amplification error:  648\n",
      "\t\tNumber of amplification errors:  14\n",
      "\t\tNumber of fragments with amplification error:  13\n",
      "\t\tNumber of fragments with amplification error (inherited from another fragment):  129\n",
      "\t\tNumber of fragments originated from other fragments:  732\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.7905709743499756\n",
      "\n",
      "\tNew number of fragments (reads):  3450\n",
      "\t\tNumber of masked fragments:  2937\n",
      "\n",
      "\tTotal time for cell's cutFragments:  0.1434478759765625\n",
      "\n",
      "\tTotal number of bases in reads:  345000\n",
      "\tTotal number of bases with amplification error:  842\n",
      "\tAmplification error per read base ratio:  0.0024405797101449277\n",
      "\n",
      "\tApproximate cell coverage:  34.5\n",
      "\n",
      "\tTotal time for cell's ampErrorCount:  0.017508983612060547\n",
      "\n",
      "\tTotal time for cell's generateReads:  16.809695720672607 4.446903944015503 12.362791776657104 11.207505464553833\n",
      "\n",
      "\tTotal time generateReads subsets:  4.3480987548828125 0.09284520149230957\n",
      "\n",
      "\tNumber of sequencing errors:  95\n",
      "\n",
      "\tTotal time for cell's generateReads:  16.82461404800415\n",
      "\tRunning:  samtools view -h -b -S ../data/2020_07_09_1/cell_idx_0.sam > ../data/2020_07_09_1/cell_idx_0.bam\n",
      "\tRunning:  samtools index -b ../data/2020_07_09_1/cell_idx_0.bam\n",
      "\n",
      "\tTotal time for cell's generateCellSam:  0.2690699100494385\n",
      "\n",
      "\tTotal time for cell :  18.04891586303711\n",
      "\n",
      "\tSimulating cell:  1\n",
      "\n",
      "\tNumber of Mutations:  70\n",
      "\n",
      "\tTotal time for cell's generateCellGenotype:  0.0007460117340087891\n",
      "\tp_ado:  0.3\n",
      "\tNumber of mask events:  54\n",
      "\t\tNumber of covered basepairs:  5286\n",
      "\t\tMask information:  {77: 1, 82: 1, 84: 2, 85: 1, 86: 2, 88: 5, 89: 1, 92: 2, 93: 2, 94: 3, 95: 1, 96: 3, 97: 2, 98: 4, 99: 3, 101: 3, 102: 2, 103: 2, 104: 1, 106: 4, 107: 1, 109: 2, 111: 1, 112: 2, 113: 1, 115: 1, 126: 1}\n",
      "\n",
      "\tTotal time for cell's maskGenome:  0.0022308826446533203\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.9740562438964844 0.12935614585876465 0.8447000980377197 0.8309950828552246\n",
      "\tNumber of rejected fragments due to amplification wave:  6844\n",
      "\n",
      "\tTotal number of fragments:  796\n",
      "\t\tNumber of fragments without amplification error:  645\n",
      "\t\tNumber of amplification errors:  8\n",
      "\t\tNumber of fragments with amplification error:  8\n",
      "\t\tNumber of fragments with amplification error (inherited from another fragment):  143\n",
      "\t\tNumber of fragments originated from other fragments:  739\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  1.0332880020141602\n",
      "\n",
      "\tNew number of fragments (reads):  4172\n",
      "\t\tNumber of masked fragments:  3262\n",
      "\n",
      "\tTotal time for cell's cutFragments:  0.17982792854309082\n",
      "\n",
      "\tTotal number of bases in reads:  417200\n",
      "\tTotal number of bases with amplification error:  614\n",
      "\tAmplification error per read base ratio:  0.0014717162032598275\n",
      "\n",
      "\tApproximate cell coverage:  41.72\n",
      "\n",
      "\tTotal time for cell's ampErrorCount:  0.018758058547973633\n",
      "\n",
      "\tTotal time for cell's generateReads:  19.481704711914062 5.13078498840332 14.350919723510742 13.007867097854614\n",
      "\n",
      "\tTotal time generateReads subsets:  5.0240559577941895 0.1004633903503418\n",
      "\n",
      "\tNumber of sequencing errors:  120\n",
      "\n",
      "\tTotal time for cell's generateReads:  19.507694959640503\n",
      "\tRunning:  samtools view -h -b -S ../data/2020_07_09_1/cell_idx_1.sam > ../data/2020_07_09_1/cell_idx_1.bam\n",
      "\tRunning:  samtools index -b ../data/2020_07_09_1/cell_idx_1.bam\n",
      "\n",
      "\tTotal time for cell's generateCellSam:  0.31026411056518555\n",
      "\n",
      "\tTotal time for cell :  21.052937030792236\n",
      "\n",
      "\tSimulating cell:  2\n",
      "\n",
      "\tNumber of Mutations:  58\n",
      "\n",
      "\tTotal time for cell's generateCellGenotype:  0.00061798095703125\n",
      "\tp_ado:  0.3\n",
      "\tNumber of mask events:  65\n",
      "\t\tNumber of covered basepairs:  6577\n",
      "\t\tMask information:  {136: 1, 79: 1, 80: 2, 82: 1, 86: 2, 87: 1, 90: 1, 91: 2, 92: 3, 93: 4, 94: 2, 95: 3, 96: 4, 97: 2, 98: 1, 99: 1, 100: 3, 101: 1, 103: 3, 104: 4, 105: 2, 106: 1, 107: 3, 109: 3, 110: 2, 111: 2, 112: 1, 113: 1, 114: 2, 116: 2, 117: 1, 118: 2, 123: 1}\n",
      "\n",
      "\tTotal time for cell's maskGenome:  0.0019087791442871094\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.6903414726257324 0.10802531242370605 0.5823161602020264 0.5701215267181396\n",
      "\tNumber of rejected fragments due to amplification wave:  6317\n",
      "\n",
      "\tTotal number of fragments:  793\n",
      "\t\tNumber of fragments without amplification error:  781\n",
      "\t\tNumber of amplification errors:  7\n",
      "\t\tNumber of fragments with amplification error:  7\n",
      "\t\tNumber of fragments with amplification error (inherited from another fragment):  5\n",
      "\t\tNumber of fragments originated from other fragments:  726\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.7414209842681885\n",
      "\n",
      "\tNew number of fragments (reads):  2840\n",
      "\t\tNumber of masked fragments:  3344\n",
      "\n",
      "\tTotal time for cell's cutFragments:  0.12285304069519043\n",
      "\n",
      "\tTotal number of bases in reads:  284000\n",
      "\tTotal number of bases with amplification error:  248\n",
      "\tAmplification error per read base ratio:  0.0008732394366197183\n",
      "\n",
      "\tApproximate cell coverage:  28.4\n",
      "\n",
      "\tTotal time for cell's ampErrorCount:  0.013261079788208008\n",
      "\n",
      "\tTotal time for cell's generateReads:  13.062949895858765 3.390803575515747 9.672146320343018 8.768839359283447\n",
      "\n",
      "\tTotal time generateReads subsets:  3.3204405307769775 0.06622815132141113\n",
      "\n",
      "\tNumber of sequencing errors:  94\n",
      "\n",
      "\tTotal time for cell's generateReads:  13.089349031448364\n",
      "\tRunning:  samtools view -h -b -S ../data/2020_07_09_1/cell_idx_2.sam > ../data/2020_07_09_1/cell_idx_2.bam\n",
      "\tRunning:  samtools index -b ../data/2020_07_09_1/cell_idx_2.bam\n",
      "\n",
      "\tTotal time for cell's generateCellSam:  0.22436904907226562\n",
      "\n",
      "\tTotal time for cell :  14.194391965866089\n",
      "\n",
      "\tSimulating cell:  3\n",
      "\n",
      "\tNumber of Mutations:  95\n",
      "\n",
      "\tTotal time for cell's generateCellGenotype:  0.0009639263153076172\n",
      "\tp_ado:  0.3\n",
      "\tNumber of mask events:  58\n",
      "\t\tNumber of covered basepairs:  6007\n",
      "\t\tMask information:  {86: 1, 89: 1, 90: 1, 91: 1, 92: 1, 93: 2, 94: 2, 95: 1, 96: 3, 97: 1, 98: 2, 99: 2, 100: 3, 101: 4, 102: 4, 103: 1, 104: 5, 105: 3, 106: 2, 107: 2, 108: 2, 109: 3, 111: 3, 112: 1, 113: 1, 115: 1, 119: 1, 124: 1, 125: 2, 126: 1}\n",
      "\n",
      "\tTotal time for cell's maskGenome:  0.003264904022216797\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.7724339962005615 0.10825538635253906 0.6641786098480225 0.6532425880432129\n",
      "\tNumber of rejected fragments due to amplification wave:  6676\n",
      "\n",
      "\tTotal number of fragments:  779\n",
      "\t\tNumber of fragments without amplification error:  650\n",
      "\t\tNumber of amplification errors:  16\n",
      "\t\tNumber of fragments with amplification error:  16\n",
      "\t\tNumber of fragments with amplification error (inherited from another fragment):  113\n",
      "\t\tNumber of fragments originated from other fragments:  718\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.824030876159668\n",
      "\n",
      "\tNew number of fragments (reads):  3507\n",
      "\t\tNumber of masked fragments:  3382\n",
      "\n",
      "\tTotal time for cell's cutFragments:  0.14607691764831543\n",
      "\n",
      "\tTotal number of bases in reads:  350700\n",
      "\tTotal number of bases with amplification error:  1138\n",
      "\tAmplification error per read base ratio:  0.0032449386940404906\n",
      "\n",
      "\tApproximate cell coverage:  35.07\n",
      "\n",
      "\tTotal time for cell's ampErrorCount:  0.014528036117553711\n",
      "\n",
      "\tTotal time for cell's generateReads:  15.742218971252441 4.139397859573364 11.602821111679077 10.517514705657959\n",
      "\n",
      "\tTotal time generateReads subsets:  4.052123069763184 0.08199667930603027\n",
      "\n",
      "\tNumber of sequencing errors:  102\n",
      "\n",
      "\tTotal time for cell's generateReads:  15.763895988464355\n",
      "\tRunning:  samtools view -h -b -S ../data/2020_07_09_1/cell_idx_3.sam > ../data/2020_07_09_1/cell_idx_3.bam\n",
      "\tRunning:  samtools index -b ../data/2020_07_09_1/cell_idx_3.bam\n",
      "\n",
      "\tTotal time for cell's generateCellSam:  0.3130459785461426\n",
      "\n",
      "\tTotal time for cell :  17.066330909729004\n",
      "\n",
      "\tSimulating cell:  4\n",
      "\n",
      "\tNumber of Mutations:  85\n",
      "\n",
      "\tTotal time for cell's generateCellGenotype:  0.0006968975067138672\n",
      "\tp_ado:  0.3\n",
      "\tNumber of mask events:  63\n",
      "\t\tNumber of covered basepairs:  6340\n",
      "\t\tMask information:  {131: 1, 80: 1, 82: 1, 84: 1, 85: 1, 88: 3, 90: 3, 92: 1, 94: 2, 95: 2, 96: 4, 97: 5, 98: 2, 100: 1, 101: 5, 102: 6, 104: 2, 105: 4, 106: 4, 107: 2, 108: 2, 109: 3, 111: 3, 112: 2, 113: 2}\n",
      "\n",
      "\tTotal time for cell's maskGenome:  0.0024640560150146484\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.7622957229614258 0.1214287281036377 0.6408669948577881 0.6278853416442871\n",
      "\tNumber of rejected fragments due to amplification wave:  7300\n",
      "\n",
      "\tTotal number of fragments:  772\n",
      "\t\tNumber of fragments without amplification error:  522\n",
      "\t\tNumber of amplification errors:  14\n",
      "\t\tNumber of fragments with amplification error:  11\n",
      "\t\tNumber of fragments with amplification error (inherited from another fragment):  239\n",
      "\t\tNumber of fragments originated from other fragments:  707\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.818213939666748\n",
      "\n",
      "\tNew number of fragments (reads):  3213\n",
      "\t\tNumber of masked fragments:  3353\n",
      "\n",
      "\tTotal time for cell's cutFragments:  0.17349886894226074\n",
      "\n",
      "\tTotal number of bases in reads:  321300\n",
      "\tTotal number of bases with amplification error:  415\n",
      "\tAmplification error per read base ratio:  0.0012916277622159974\n",
      "\n",
      "\tApproximate cell coverage:  32.13\n",
      "\n",
      "\tTotal time for cell's ampErrorCount:  0.01741814613342285\n",
      "\n",
      "\tTotal time for cell's generateReads:  14.265105247497559 3.728207588195801 10.536897659301758 9.556692600250244\n",
      "\n",
      "\tTotal time generateReads subsets:  3.651777982711792 0.07168459892272949\n",
      "\n",
      "\tNumber of sequencing errors:  109\n",
      "\n",
      "\tTotal time for cell's generateReads:  14.288003921508789\n",
      "\tRunning:  samtools view -h -b -S ../data/2020_07_09_1/cell_idx_4.sam > ../data/2020_07_09_1/cell_idx_4.bam\n",
      "\tRunning:  samtools index -b ../data/2020_07_09_1/cell_idx_4.bam\n",
      "\n",
      "\tTotal time for cell's generateCellSam:  0.2781679630279541\n",
      "\n",
      "\tTotal time for cell :  15.578624963760376\n",
      "\n",
      "All cell totals\n",
      "\tTotal number of bases in reads:  1718200\n",
      "\tTotal number of bases with amplification error:  3257\n",
      "\tAmplification error per read base ratio:  0.0018955884064718893\n",
      "\n",
      "\tApproximate average cell coverages:  34.364\n",
      "\n",
      "Total time:  85.94129800796509\n",
      "Part 4 ends...\n",
      "\n",
      "Part 5: Relocate files\n",
      "\n",
      "\tDirectory is created:  ../data/2020_07_09_1/bam/\n",
      "\tDirectory is created:  ../data/2020_07_09_1/truth/\n",
      "\n",
      "Total time:  0.06576991081237793\n",
      "Part 5 ends...\n",
      "\n",
      "Total (global) time:  87.4329559803009\n",
      "Data simulation ends...\n"
     ]
    }
   ],
   "source": [
    "%run data_simulator.py --global_dir \"../data/2020_07_09_1/\" --genome_length 10000 --seed_val 42 --num_cells 5 --amp_method mda --num_iter 1000 --p_ado 0.3 --ado_poisson_rate 100 --ado_type 3 --p_ae 0.00002"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
