{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 05. Pipeline Script Generator\n",
    "\n",
    "This notebook automatically generates needed scripts for the cluster. The scripts are Uppmax-compatible, it loads modules such as bioinfo-tools, pysam, samtools etc. Further modifications are needed to be done for Tetralith compatibility (such as activating a Conda environment).  \n",
    "\n",
    "Steps: \n",
    "* Generate scripts (pipeline_scripts.py)\n",
    "\n",
    "For Tetralith:\n",
    "* Remove #SBATCH -p core line\n",
    "* Remove module load python3\n",
    "* Add module load Python/3.7.0-anaconda-5.3.0-extras-nsc1\n",
    "* Add source activate /proj/sc_ml/users/x_hazko/scuphr_conda/scuphr_env (or another conda environment)\n",
    "* Also, the number of allocated cores can be increased."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: pipeline_scripts.py [-h] [--chr_id CHR_ID] [--data_type DATA_TYPE]\n",
      "                           [--global_dir GLOBAL_DIR] [--num_cells NUM_CELLS]\n",
      "                           [--output_dir OUTPUT_DIR] [--p_ado P_ADO]\n",
      "                           [--p_ae P_AE] [--pos_range_min POS_RANGE_MIN]\n",
      "                           [--pos_range_max POS_RANGE_MAX]\n",
      "                           [--print_status PRINT_STATUS]\n",
      "                           [--read_length READ_LENGTH] [--seed_val SEED_VAL]\n",
      "                           [--ado_poisson_rate ADO_POISSON_RATE]\n",
      "                           [--ado_type ADO_TYPE] [--amp_method AMP_METHOD]\n",
      "                           [--genome_length GENOME_LENGTH]\n",
      "                           [--gsnv_rate GSNV_RATE] [--is_flat IS_FLAT]\n",
      "                           [--mut_poisson_rate MUT_POISSON_RATE]\n",
      "                           [--num_iter NUM_ITER]\n",
      "                           [--num_max_mid_points NUM_MAX_MID_POINTS]\n",
      "                           [--num_rep_amp_bias NUM_REP_AMP_BIAS]\n",
      "                           [--phred_type PHRED_TYPE]\n",
      "                           [--bulk_depth_threshold BULK_DEPTH_THRESHOLD]\n",
      "                           [--cell_depth_threshold CELL_DEPTH_THRESHOLD]\n",
      "                           [--het_ratio_threshold HET_RATIO_THRESHOLD]\n",
      "                           [--min_line MIN_LINE] [--max_line MAX_LINE]\n",
      "                           [--nuc_depth_threshold NUC_DEPTH_THRESHOLD]\n",
      "                           [--min_read_count MIN_READ_COUNT]\n",
      "                           [--max_read_count MAX_READ_COUNT]\n",
      "                           [--min_cell_count MIN_CELL_COUNT]\n",
      "                           [--output_dict_dir OUTPUT_DICT_DIR] [--a_g A_G]\n",
      "                           [--b_g B_G]\n",
      "                           project_name project_dir\n",
      "\n",
      "Generate scripts for the pipeline.\n",
      "\n",
      "positional arguments:\n",
      "  project_name          Specify the project name. Needed for -A tag in bash\n",
      "                        script.\n",
      "  project_dir           Specify the source code directory, i.e. /proj/scuphr/\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  --chr_id CHR_ID       Specify the chromosome number. Default: 1\n",
      "  --data_type DATA_TYPE\n",
      "                        Specify the data type. Default: real\n",
      "  --global_dir GLOBAL_DIR\n",
      "                        Specify the directory.\n",
      "  --num_cells NUM_CELLS\n",
      "                        Specify the number of cells. Default: 10\n",
      "  --output_dir OUTPUT_DIR\n",
      "                        Specify the output directory.\n",
      "  --p_ado P_ADO         Specify the allelic dropout probability of a base.\n",
      "                        Default: 0.005\n",
      "  --p_ae P_AE           Specify the amplification error probability of a base.\n",
      "                        Default: 0.00001\n",
      "  --pos_range_min POS_RANGE_MIN\n",
      "                        Specify the position range (min value).\n",
      "  --pos_range_max POS_RANGE_MAX\n",
      "                        Specify the position range (max value).\n",
      "  --print_status PRINT_STATUS\n",
      "                        Specify the print (0 for do not print, 1 for print).\n",
      "                        Default: 0\n",
      "  --read_length READ_LENGTH\n",
      "                        Specify the read length. Default: 100\n",
      "  --seed_val SEED_VAL   Specify the seed. Default: 123\n",
      "  --ado_poisson_rate ADO_POISSON_RATE\n",
      "                        Specify the ado poisson rate. Default: 0\n",
      "  --ado_type ADO_TYPE   Specify the allelic dropout type (0 for no ado, 1 for\n",
      "                        random Beta, 2 for even smaller Beta, 3 for fixed.\n",
      "                        Default: 3).\n",
      "  --amp_method AMP_METHOD\n",
      "                        Specify the amplification method. Default: mda\n",
      "  --genome_length GENOME_LENGTH\n",
      "                        Specify the length of genome. Default: 50000\n",
      "  --gsnv_rate GSNV_RATE\n",
      "                        Specify the gSNV rate. Default: 0.005\n",
      "  --is_flat IS_FLAT     Specify the amplification bias type (True for flat,\n",
      "                        False for sine wave). Default: False\n",
      "  --mut_poisson_rate MUT_POISSON_RATE\n",
      "                        Specify the mutation poisson rate. Default: 3\n",
      "  --num_iter NUM_ITER   Specify the number of iteration. Default: 5000\n",
      "  --num_max_mid_points NUM_MAX_MID_POINTS\n",
      "                        Specify the number of division points of amplification\n",
      "                        bias. Default: 10\n",
      "  --num_rep_amp_bias NUM_REP_AMP_BIAS\n",
      "                        Specify the number of repetitions of amplification\n",
      "                        bias. Default: 3\n",
      "  --phred_type PHRED_TYPE\n",
      "                        Specify the phred score type (0 for no errors, 1 for\n",
      "                        all 42, 2 for truncated normal). Default: 2\n",
      "  --bulk_depth_threshold BULK_DEPTH_THRESHOLD\n",
      "                        Specify the bulk depth threshold. Default: 20\n",
      "  --cell_depth_threshold CELL_DEPTH_THRESHOLD\n",
      "                        Specify the cell depth threshold. Default: 0\n",
      "  --het_ratio_threshold HET_RATIO_THRESHOLD\n",
      "                        Specify the bulk heterozygous ratio threshold.\n",
      "                        Default: 0.2\n",
      "  --min_line MIN_LINE   Specify the line number of min het position. Default:\n",
      "                        0\n",
      "  --max_line MAX_LINE   Specify the line number of max het position. Default:\n",
      "                        0\n",
      "  --nuc_depth_threshold NUC_DEPTH_THRESHOLD\n",
      "                        Specify the minimum number of valid reads. Default: 2\n",
      "  --min_read_count MIN_READ_COUNT\n",
      "                        Specify the minimum read count. Default: 0\n",
      "  --max_read_count MAX_READ_COUNT\n",
      "                        Specify the maximum read count (0 for all). Default: 0\n",
      "  --min_cell_count MIN_CELL_COUNT\n",
      "                        Specify the minimum cell count. Default: 2\n",
      "  --output_dict_dir OUTPUT_DICT_DIR\n",
      "                        Specify the output dictionary directory.\n",
      "  --a_g A_G             Specify the alpha prior of mutation probability.\n",
      "                        Default: 1\n",
      "  --b_g B_G             Specify the beta prior of mutation probability.\n",
      "                        Default: 1\n"
     ]
    }
   ],
   "source": [
    "%run pipeline_scripts.py --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Create necessary folders\n",
      "\n",
      "Get all parameters\n",
      "\n",
      "Writing data simulation script\n",
      "\n",
      "Script is saved to  ../data/example_dataset/scripts/script_01_data_simulation.sh\n",
      "\n",
      "Writing site detection par script\n",
      "\n",
      "Script is saved to  ../data/example_dataset/scripts/script_02_site_detection.sh\n",
      "\n",
      "Writing generate dataset script\n",
      "\n",
      "Script is saved to  ../data/example_dataset/scripts/script_03_generate_dataset.sh\n",
      "\n",
      "Writing distance between cells script\n",
      "\n",
      "Script is saved to  ../results/example_results/scripts/script_04_dbc.sh\n",
      "\n",
      "Writing distance between cells combine script\n",
      "\n",
      "Script is saved to  ../results/example_results/scripts/script_05_dbc_combine.sh\n",
      "\n",
      "Writing lineage tree script\n",
      "\n",
      "Script is saved to  ../results/example_results/scripts/script_06_lineage_tree.sh\n",
      "\n",
      "Writing combined data par simulation script\n",
      "\n",
      "Script is saved to  ../data/example_dataset/scripts/script_00_combined_data_par.sh\n"
     ]
    }
   ],
   "source": [
    "%run pipeline_scripts.py snic2020-5-278 ../ --num_cells 5 --genome_length 20000 --data_type synthetic --global_dir example_dataset/ --output_dir example_results/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Example Usage\n",
    "\n",
    "Below is the pipeline script generator. I generated a small dataset. Then, in the following cells in this notebook, I ran the generated commands in the scripts one by one. Normally, one needs to submit jobs in the cluster one by one. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Create necessary folders\n",
      "\n",
      "Get all parameters\n",
      "\n",
      "Writing data simulation script\n",
      "\n",
      "Script is saved to  ../data/2020_07_09/scripts/script_01_data_simulation.sh\n",
      "\n",
      "Writing site detection par script\n",
      "\n",
      "Script is saved to  ../data/2020_07_09/scripts/script_02_site_detection.sh\n",
      "\n",
      "Writing generate dataset script\n",
      "\n",
      "Script is saved to  ../data/2020_07_09/scripts/script_03_generate_dataset.sh\n",
      "\n",
      "Writing distance between cells script\n",
      "\n",
      "Script is saved to  ../results/2020_07_09_results/scripts/script_04_dbc.sh\n",
      "\n",
      "Writing distance between cells combine script\n",
      "\n",
      "Script is saved to  ../results/2020_07_09_results/scripts/script_05_dbc_combine.sh\n",
      "\n",
      "Writing lineage tree script\n",
      "\n",
      "Script is saved to  ../results/2020_07_09_results/scripts/script_06_lineage_tree.sh\n",
      "\n",
      "Writing combined data par simulation script\n",
      "\n",
      "Script is saved to  ../data/2020_07_09/scripts/script_00_combined_data_par.sh\n"
     ]
    }
   ],
   "source": [
    "%run pipeline_scripts.py snic2020-5-278 ../ --data_type synthetic --global_dir 2020_07_09/ --genome_length 10000 --seed_val 42 --num_cells 5 --num_iter 1000 --p_ado 0.3 --ado_poisson_rate 100 --p_ae 0.00002 --max_read_count 10 --output_dir 2020_07_09_results/   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simulate Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The data directory:  ../data/2020_07_09/\n",
      "\n",
      "Part 0: Save arguments\n",
      "\n",
      "\n",
      "Total time:  0.0007112026214599609\n",
      "Part 0 ends...\n",
      "\n",
      "Data simulation starts...\n",
      "\n",
      "Part 1: Amplification bias\n",
      "\n",
      "\tMiddle points for repetition  0 :  [ 466  860 4426 5191 5390 5734 6265]\n",
      "\tMiddle points for repetition  1 :  [2064 2303 3392 7985 8499]\n",
      "\tMiddle points for repetition  2 :  [ 571 2144 3491 3971 9389]\n",
      "\n",
      "Total time:  0.37949180603027344\n",
      "Part 1 ends...\n",
      "\n",
      "Part 2: Reference, bulk and mutation generation\n",
      "\n",
      "\tnumBp:  10000\n",
      "\tgSNV count:  50\n",
      "\tmutation count:  250\n",
      "\tTotal gSNV count:  50 \tTotal mutation count:  250 \tMutations close to a gSNV:  126\n",
      "\tRunning:  samtools view ../data/2020_07_09/bulk_ref.sam | awk '{OFS=\"\\t\"; print \">\"$1\"\\n\"$10}' - > ../data/2020_07_09/bulk_ref.fasta\n",
      "\tRunning:  samtools view -h -b -S ../data/2020_07_09/bulk.sam > ../data/2020_07_09/bulk.bam\n",
      "\tRunning:  samtools index -b ../data/2020_07_09/bulk.bam\n",
      "\n",
      "Total time:  1.726686954498291\n",
      "Part 1 ends...\n",
      "\n",
      "Part 3: Tree generation and mutation partition\n",
      "\n",
      "\tparent nodes:  [0 0 0 1 1 2 2 5 5]\n",
      "\tleaf nodes:  [3, 4, 6, 7, 8]\n",
      "\n",
      "Total time:  0.0024771690368652344\n",
      "Part 3 ends...\n",
      "\n",
      "Part 4: Cell simulation\n",
      "\n",
      "\n",
      "\tSimulating cell:  0\n",
      "\n",
      "\tNumber of Mutations:  58\n",
      "\n",
      "\tTotal time for cell's generateCellGenotype:  0.0009431838989257812\n",
      "\tp_ado:  0.3\n",
      "\tNumber of mask events:  63\n",
      "\t\tNumber of covered basepairs:  6327\n",
      "\t\tMask information:  {128: 1, 80: 1, 56: 1, 83: 2, 85: 2, 86: 1, 87: 1, 88: 2, 90: 1, 91: 1, 92: 5, 93: 2, 94: 2, 95: 2, 96: 1, 99: 2, 100: 3, 101: 2, 102: 4, 104: 7, 105: 1, 106: 2, 107: 1, 109: 3, 110: 1, 111: 3, 112: 1, 113: 1, 115: 1, 116: 1, 117: 2, 118: 1, 119: 1, 120: 1}\n",
      "\n",
      "\tTotal time for cell's maskGenome:  0.004021883010864258\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.9362664222717285 0.1517953872680664 0.7844710350036621 0.7671377658843994\n",
      "\tNumber of rejected fragments due to amplification wave:  6470\n",
      "\n",
      "\tTotal number of fragments:  790\n",
      "\t\tNumber of fragments without amplification error:  648\n",
      "\t\tNumber of amplification errors:  14\n",
      "\t\tNumber of fragments with amplification error:  13\n",
      "\t\tNumber of fragments with amplification error (inherited from another fragment):  129\n",
      "\t\tNumber of fragments originated from other fragments:  732\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  1.0116760730743408\n",
      "\n",
      "\tNew number of fragments (reads):  3450\n",
      "\t\tNumber of masked fragments:  2937\n",
      "\n",
      "\tTotal time for cell's cutFragments:  0.1982588768005371\n",
      "\n",
      "\tTotal number of bases in reads:  345000\n",
      "\tTotal number of bases with amplification error:  842\n",
      "\tAmplification error per read base ratio:  0.0024405797101449277\n",
      "\n",
      "\tApproximate cell coverage:  34.5\n",
      "\n",
      "\tTotal time for cell's ampErrorCount:  0.021262168884277344\n",
      "\n",
      "\tTotal time for cell's generateReads:  16.229557037353516 4.2105937004089355 12.01896333694458 10.93556523323059\n",
      "\n",
      "\tTotal time generateReads subsets:  4.124113321304321 0.08089065551757812\n",
      "\n",
      "\tNumber of sequencing errors:  95\n",
      "\n",
      "\tTotal time for cell's generateReads:  16.242316961288452\n",
      "\tRunning:  samtools view -h -b -S ../data/2020_07_09/cell_idx_0.sam > ../data/2020_07_09/cell_idx_0.bam\n",
      "\tRunning:  samtools index -b ../data/2020_07_09/cell_idx_0.bam\n",
      "\n",
      "\tTotal time for cell's generateCellSam:  0.263059139251709\n",
      "\n",
      "\tTotal time for cell :  17.742071866989136\n",
      "\n",
      "\tSimulating cell:  1\n",
      "\n",
      "\tNumber of Mutations:  70\n",
      "\n",
      "\tTotal time for cell's generateCellGenotype:  0.0006117820739746094\n",
      "\tp_ado:  0.3\n",
      "\tNumber of mask events:  54\n",
      "\t\tNumber of covered basepairs:  5286\n",
      "\t\tMask information:  {77: 1, 82: 1, 84: 2, 85: 1, 86: 2, 88: 5, 89: 1, 92: 2, 93: 2, 94: 3, 95: 1, 96: 3, 97: 2, 98: 4, 99: 3, 101: 3, 102: 2, 103: 2, 104: 1, 106: 4, 107: 1, 109: 2, 111: 1, 112: 2, 113: 1, 115: 1, 126: 1}\n",
      "\n",
      "\tTotal time for cell's maskGenome:  0.0022537708282470703\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.8528659343719482 0.11234140396118164 0.7405245304107666 0.7290067672729492\n",
      "\tNumber of rejected fragments due to amplification wave:  6844\n",
      "\n",
      "\tTotal number of fragments:  796\n",
      "\t\tNumber of fragments without amplification error:  645\n",
      "\t\tNumber of amplification errors:  8\n",
      "\t\tNumber of fragments with amplification error:  8\n",
      "\t\tNumber of fragments with amplification error (inherited from another fragment):  143\n",
      "\t\tNumber of fragments originated from other fragments:  739\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.9069831371307373\n",
      "\n",
      "\tNew number of fragments (reads):  4172\n",
      "\t\tNumber of masked fragments:  3262\n",
      "\n",
      "\tTotal time for cell's cutFragments:  0.1870419979095459\n",
      "\n",
      "\tTotal number of bases in reads:  417200\n",
      "\tTotal number of bases with amplification error:  614\n",
      "\tAmplification error per read base ratio:  0.0014717162032598275\n",
      "\n",
      "\tApproximate cell coverage:  41.72\n",
      "\n",
      "\tTotal time for cell's ampErrorCount:  0.0230410099029541\n",
      "\n",
      "\tTotal time for cell's generateReads:  17.595046043395996 4.5717620849609375 13.023283958435059 11.881494998931885\n",
      "\n",
      "\tTotal time generateReads subsets:  4.4829864501953125 0.08363699913024902\n",
      "\n",
      "\tNumber of sequencing errors:  120\n",
      "\n",
      "\tTotal time for cell's generateReads:  17.620057821273804\n",
      "\tRunning:  samtools view -h -b -S ../data/2020_07_09/cell_idx_1.sam > ../data/2020_07_09/cell_idx_1.bam\n",
      "\tRunning:  samtools index -b ../data/2020_07_09/cell_idx_1.bam\n",
      "\n",
      "\tTotal time for cell's generateCellSam:  0.305042028427124\n",
      "\n",
      "\tTotal time for cell :  19.045221090316772\n",
      "\n",
      "\tSimulating cell:  2\n",
      "\n",
      "\tNumber of Mutations:  58\n",
      "\n",
      "\tTotal time for cell's generateCellGenotype:  0.0006129741668701172\n",
      "\tp_ado:  0.3\n",
      "\tNumber of mask events:  65\n",
      "\t\tNumber of covered basepairs:  6577\n",
      "\t\tMask information:  {136: 1, 79: 1, 80: 2, 82: 1, 86: 2, 87: 1, 90: 1, 91: 2, 92: 3, 93: 4, 94: 2, 95: 3, 96: 4, 97: 2, 98: 1, 99: 1, 100: 3, 101: 1, 103: 3, 104: 4, 105: 2, 106: 1, 107: 3, 109: 3, 110: 2, 111: 2, 112: 1, 113: 1, 114: 2, 116: 2, 117: 1, 118: 2, 123: 1}\n",
      "\n",
      "\tTotal time for cell's maskGenome:  0.0018229484558105469\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.652517557144165 0.09779524803161621 0.5547223091125488 0.5444943904876709\n",
      "\tNumber of rejected fragments due to amplification wave:  6317\n",
      "\n",
      "\tTotal number of fragments:  793\n",
      "\t\tNumber of fragments without amplification error:  781\n",
      "\t\tNumber of amplification errors:  7\n",
      "\t\tNumber of fragments with amplification error:  7\n",
      "\t\tNumber of fragments with amplification error (inherited from another fragment):  5\n",
      "\t\tNumber of fragments originated from other fragments:  726\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.6999809741973877\n",
      "\n",
      "\tNew number of fragments (reads):  2840\n",
      "\t\tNumber of masked fragments:  3344\n",
      "\n",
      "\tTotal time for cell's cutFragments:  0.12699580192565918\n",
      "\n",
      "\tTotal number of bases in reads:  284000\n",
      "\tTotal number of bases with amplification error:  248\n",
      "\tAmplification error per read base ratio:  0.0008732394366197183\n",
      "\n",
      "\tApproximate cell coverage:  28.4\n",
      "\n",
      "\tTotal time for cell's ampErrorCount:  0.01710796356201172\n",
      "\n",
      "\tTotal time for cell's generateReads:  14.555234670639038 3.823009729385376 10.732224941253662 9.747466087341309\n",
      "\n",
      "\tTotal time generateReads subsets:  3.734945297241211 0.08354735374450684\n",
      "\n",
      "\tNumber of sequencing errors:  94\n",
      "\n",
      "\tTotal time for cell's generateReads:  14.584752082824707\n",
      "\tRunning:  samtools view -h -b -S ../data/2020_07_09/cell_idx_2.sam > ../data/2020_07_09/cell_idx_2.bam\n",
      "\tRunning:  samtools index -b ../data/2020_07_09/cell_idx_2.bam\n",
      "\n",
      "\tTotal time for cell's generateCellSam:  0.3115711212158203\n",
      "\n",
      "\tTotal time for cell :  15.743005990982056\n",
      "\n",
      "\tSimulating cell:  3\n",
      "\n",
      "\tNumber of Mutations:  95\n",
      "\n",
      "\tTotal time for cell's generateCellGenotype:  0.0006299018859863281\n",
      "\tp_ado:  0.3\n",
      "\tNumber of mask events:  58\n",
      "\t\tNumber of covered basepairs:  6007\n",
      "\t\tMask information:  {86: 1, 89: 1, 90: 1, 91: 1, 92: 1, 93: 2, 94: 2, 95: 1, 96: 3, 97: 1, 98: 2, 99: 2, 100: 3, 101: 4, 102: 4, 103: 1, 104: 5, 105: 3, 106: 2, 107: 2, 108: 2, 109: 3, 111: 3, 112: 1, 113: 1, 115: 1, 119: 1, 124: 1, 125: 2, 126: 1}\n",
      "\n",
      "\tTotal time for cell's maskGenome:  0.0021500587463378906\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.9203774929046631 0.19464492797851562 0.7257325649261475 0.7114315032958984\n",
      "\tNumber of rejected fragments due to amplification wave:  6676\n",
      "\n",
      "\tTotal number of fragments:  779\n",
      "\t\tNumber of fragments without amplification error:  650\n",
      "\t\tNumber of amplification errors:  16\n",
      "\t\tNumber of fragments with amplification error:  16\n",
      "\t\tNumber of fragments with amplification error (inherited from another fragment):  113\n",
      "\t\tNumber of fragments originated from other fragments:  718\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.9905338287353516\n",
      "\n",
      "\tNew number of fragments (reads):  3507\n",
      "\t\tNumber of masked fragments:  3382\n",
      "\n",
      "\tTotal time for cell's cutFragments:  0.22684597969055176\n",
      "\n",
      "\tTotal number of bases in reads:  350700\n",
      "\tTotal number of bases with amplification error:  1138\n",
      "\tAmplification error per read base ratio:  0.0032449386940404906\n",
      "\n",
      "\tApproximate cell coverage:  35.07\n",
      "\n",
      "\tTotal time for cell's ampErrorCount:  0.019436120986938477\n",
      "\n",
      "\tTotal time for cell's generateReads:  17.07466220855713 4.435115575790405 12.639546632766724 11.47968602180481\n",
      "\n",
      "\tTotal time generateReads subsets:  4.340879917144775 0.08880877494812012\n",
      "\n",
      "\tNumber of sequencing errors:  102\n",
      "\n",
      "\tTotal time for cell's generateReads:  17.10039210319519\n",
      "\tRunning:  samtools view -h -b -S ../data/2020_07_09/cell_idx_3.sam > ../data/2020_07_09/cell_idx_3.bam\n",
      "\tRunning:  samtools index -b ../data/2020_07_09/cell_idx_3.bam\n",
      "\n",
      "\tTotal time for cell's generateCellSam:  0.2946779727935791\n",
      "\n",
      "\tTotal time for cell :  18.635205030441284\n",
      "\n",
      "\tSimulating cell:  4\n",
      "\n",
      "\tNumber of Mutations:  85\n",
      "\n",
      "\tTotal time for cell's generateCellGenotype:  0.0007069110870361328\n",
      "\tp_ado:  0.3\n",
      "\tNumber of mask events:  63\n",
      "\t\tNumber of covered basepairs:  6340\n",
      "\t\tMask information:  {131: 1, 80: 1, 82: 1, 84: 1, 85: 1, 88: 3, 90: 3, 92: 1, 94: 2, 95: 2, 96: 4, 97: 5, 98: 2, 100: 1, 101: 5, 102: 6, 104: 2, 105: 4, 106: 4, 107: 2, 108: 2, 109: 3, 111: 3, 112: 2, 113: 2}\n",
      "\n",
      "\tTotal time for cell's maskGenome:  0.002343893051147461\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.7442061901092529 0.1188192367553711 0.6253869533538818 0.612999439239502\n",
      "\tNumber of rejected fragments due to amplification wave:  7300\n",
      "\n",
      "\tTotal number of fragments:  772\n",
      "\t\tNumber of fragments without amplification error:  522\n",
      "\t\tNumber of amplification errors:  14\n",
      "\t\tNumber of fragments with amplification error:  11\n",
      "\t\tNumber of fragments with amplification error (inherited from another fragment):  239\n",
      "\t\tNumber of fragments originated from other fragments:  707\n",
      "\n",
      "\tTotal time for cell's amplifyGenome:  0.8008110523223877\n",
      "\n",
      "\tNew number of fragments (reads):  3213\n",
      "\t\tNumber of masked fragments:  3353\n",
      "\n",
      "\tTotal time for cell's cutFragments:  0.13899493217468262\n",
      "\n",
      "\tTotal number of bases in reads:  321300\n",
      "\tTotal number of bases with amplification error:  415\n",
      "\tAmplification error per read base ratio:  0.0012916277622159974\n",
      "\n",
      "\tApproximate cell coverage:  32.13\n",
      "\n",
      "\tTotal time for cell's ampErrorCount:  0.013173103332519531\n",
      "\n",
      "\tTotal time for cell's generateReads:  13.080731630325317 3.3785338401794434 9.702197790145874 8.848930358886719\n",
      "\n",
      "\tTotal time generateReads subsets:  3.3149046897888184 0.059851884841918945\n",
      "\n",
      "\tNumber of sequencing errors:  109\n",
      "\n",
      "\tTotal time for cell's generateReads:  13.10185980796814\n",
      "\tRunning:  samtools view -h -b -S ../data/2020_07_09/cell_idx_4.sam > ../data/2020_07_09/cell_idx_4.bam\n",
      "\tRunning:  samtools index -b ../data/2020_07_09/cell_idx_4.bam\n",
      "\n",
      "\tTotal time for cell's generateCellSam:  0.2519998550415039\n",
      "\n",
      "\tTotal time for cell :  14.310358047485352\n",
      "\n",
      "All cell totals\n",
      "\tTotal number of bases in reads:  1718200\n",
      "\tTotal number of bases with amplification error:  3257\n",
      "\tAmplification error per read base ratio:  0.0018955884064718893\n",
      "\n",
      "\tApproximate average cell coverages:  34.364\n",
      "\n",
      "Total time:  85.47597908973694\n",
      "Part 4 ends...\n",
      "\n",
      "Part 5: Relocate files\n",
      "\n",
      "\n",
      "Total time:  0.11164212226867676\n",
      "Part 5 ends...\n",
      "\n",
      "Total (global) time:  87.69863295555115\n",
      "Data simulation ends...\n"
     ]
    }
   ],
   "source": [
    "%run data_simulator.py  --seed_val 42 --p_ae 2e-05 --num_max_mid_points 10 --ado_poisson_rate 100.0 --phred_type 2 --chr_id 1 --mut_poisson_rate 3 --is_flat False --num_iter 1000 --p_ado 0.3 --amp_method mda --num_rep_amp_bias 3 --num_cells 5 --global_dir ../data/2020_07_09/ --read_length 100 --gsnv_rate 0.005 --ado_type 3 --genome_length 10000"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Detect Sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Global directory:  ../data/2020_07_09/\n",
      "\n",
      "*****STEP 1\n",
      "\n",
      "\n",
      "Pair dict is saved to:  ../data/2020_07_09/processed_data/pair_dict.pickle\n",
      "\n",
      "Pair dict is saved to:  ../data/2020_07_09/processed_data/pair_dict_0_0.pickle\n",
      "\n",
      "Number of selected pairs:  5\n",
      "\tNumber of unique het. sites:  6\n",
      "\tNumber of unique hom. sites:  5\n",
      "\tNumber of removed duplicate homozygous sites:  1\n",
      "\tTotal number of pairs:  5\n",
      "\n",
      "*****\n",
      "Total time:  158.24505305290222\n",
      "\n",
      "*****STEP 2\n",
      "\n",
      "\tBulk pairs dictionary is saved to: \t ../data/2020_07_09/processed_data/bulk_pairs.pickle\n",
      "\tCell pairs dictionary is saved to: \t ../data/2020_07_09/processed_data/cell_0_pairs.pickle\n",
      "\tCell pairs dictionary is saved to: \t ../data/2020_07_09/processed_data/cell_1_pairs.pickle\n",
      "\tCell pairs dictionary is saved to: \t ../data/2020_07_09/processed_data/cell_2_pairs.pickle\n",
      "\tCell pairs dictionary is saved to: \t ../data/2020_07_09/processed_data/cell_3_pairs.pickle\n",
      "\tCell pairs dictionary is saved to: \t ../data/2020_07_09/processed_data/cell_4_pairs.pickle\n",
      "\tAll cell pairs dictionary is saved to: \t ../data/2020_07_09/processed_data/all_cell_dict_pair.pickle\n",
      "\n",
      "*****\n",
      "Total time:  0.002534151077270508\n",
      "\n",
      "*****DONE!\n",
      "\n",
      "\n",
      "*****\n",
      "Total global time:  158.24768996238708\n"
     ]
    }
   ],
   "source": [
    "%run site_detection.py ../data/2020_07_09/ 5 --seed_val 42  --min_line 0  --print_status 0  --het_ratio_threshold 0.2  --chr_id 1  --nuc_depth_threshold 2  --max_line 0  --bulk_depth_threshold 20  --cell_depth_threshold 0  --read_length 100 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Global directory:  ../data/2020_07_09/\n",
      "Output directory:  ../data/2020_07_09/processed_data_dict/\n",
      "\n",
      "Part 1: Load bulk pairs dictionary.\n",
      "\n",
      "Total time:  0.00018787384033203125\n",
      "Part 1 ends...\n",
      "\n",
      "Part 2: Load cells' pairs dictionaries.\n",
      "\n",
      "\n",
      "Total time:  0.001049041748046875\n",
      "Part 2 ends...\n",
      "\n",
      "Part 3: Generate dataset.\n",
      "\n",
      "\tTotal number of positions:  5\n",
      "\tTotal number of skipped positions:  0\n",
      "\n",
      "Total time:  0.014261007308959961\n",
      "Part 3 ends...\n",
      "\n",
      "Part 4: Shuffle and partition the dataset.\n",
      "\n",
      "\n",
      "Shuffle and partition dataset\n",
      "\n",
      "Shuffled data ( 5 ) is saved to:  ../data/2020_07_09/processed_data_dict/data_shuffled.pickle\n",
      "Saving test data ( 2 ) is saved to:  ../data/2020_07_09/processed_data_dict/data_test.pickle\n",
      "Saving train data ( 3 ) is saved to:  ../data/2020_07_09/processed_data_dict/data_train.pickle\n",
      "Saving shuffled data ( 5 ) is saved to:  ../data/2020_07_09/processed_data_dict/data.pickle\n",
      "\n",
      "Total time:  0.004559993743896484\n",
      "Part 4 ends...\n",
      "\n",
      "Part 5: Analyse results.\n",
      "\n",
      "\tSite detection results: \n",
      "\tTotal number of (real) mutated sites:  250\n",
      "\n",
      "\tTotal number of (real) detectable mutated sites (close to gSNV locations):  126\n",
      "\t\tNumber of (real) detectable mutated sites (on internal edges): \t 44\n",
      "\t\tNumber of (real) detectable mutated sites (on leaf edges): \t 82\n",
      "\n",
      "\tTotal number of sites correctly detected \t(True Positive): \t 5\n",
      "\t\tNumber of mutated sites (detected) (on internal edges): \t 5 \t out of  44 \tperc:  11.364\n",
      "\t\tNumber of mutated sites (detected) (on leaf edges): \t\t 0 \t out of  82 \tperc:  0.0\n",
      "\tTotal number of sites incorrectly detected \t(False Positive): \t 0\n",
      "\n",
      "\tTotal number of mutated sites undetected \t(False Negative): \t 121\n",
      "\n",
      "\tSensitivity (True positive rate): \t 0.03968253968253968\n",
      "\tPrecision: \t\t\t\t 1.0\n",
      "\tMiss rate (False negative rate): \t 0.9603174603174603\n",
      "\tFalse discovery rate: \t\t\t 0.0\n",
      "\tF1 score: \t\t\t\t 0.07633587786259542\n",
      "\n",
      "\tUndetected inner edge mutations:  39\n",
      "[80, 88, 259, 416, 1494, 1509, 1593, 2020, 2233, 2344, 2592, 2802, 3393, 3410, 3898, 3912, 4003, 4049, 4081, 4442, 4747, 4751, 4822, 5323, 5741, 5970, 6261, 6442, 6517, 7146, 7417, 7438, 8181, 8446, 8846, 9238, 9335, 9412, 9445]\n",
      "\n",
      "Total time:  0.0062160491943359375\n",
      "Part 5 ends...\n",
      "\n",
      "All done!\n",
      "Total time:  0.026544809341430664\n"
     ]
    }
   ],
   "source": [
    "%run generate_dataset.py ../data/2020_07_09/ 5 --seed_val 42  --chr_id 1  --data_type synthetic  --min_read_count 0  --output_dict_dir \"\"  --max_read_count 10  --min_cell_count 2  --read_length 100 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate distance between cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "*****\n",
      "Load the dataset and read probabilities...\n",
      "\n",
      "The dataset is loaded. Number of cells: 5, number of position pairs: 5\n",
      "\n",
      "Position range: [ 0 ,  5 )\n",
      "Number of positions in range:  5\n",
      "\n",
      "*****\n",
      "Calculating real distances...\n",
      "\n",
      "*****\n",
      "Calculating real distances finished...\n",
      "\n",
      "******\n",
      "Analysing Real DBC\n",
      "******\n",
      "Analysing Real DBC\n",
      "******\n",
      "Analysing Real DBC\n",
      "******\n",
      "Analysing Real DBC\n",
      "\n",
      "\n",
      "\n",
      "Pos idx of process is:  1\n",
      "Pos idx of process is:  0\n",
      "Pos idx of process is:  3\n",
      "Pos idx of process is:  2\n",
      "Process id:  95769 . Uname:  posix.uname_result(sysname='Darwin', nodename='Hazals-MBP.lan', release='19.2.0', version='Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64', machine='x86_64')\n",
      "Process id:  95768 . Uname:  posix.uname_result(sysname='Darwin', nodename='Hazals-MBP.lan', release='19.2.0', version='Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64', machine='x86_64')\n",
      "Process id:  95771 . Uname:  posix.uname_result(sysname='Darwin', nodename='Hazals-MBP.lan', release='19.2.0', version='Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64', machine='x86_64')\n",
      "Process id:  95770 . Uname:  posix.uname_result(sysname='Darwin', nodename='Hazals-MBP.lan', release='19.2.0', version='Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64', machine='x86_64')\n",
      "\n",
      "***********\n",
      "Position: \n",
      "***********\n",
      "Position: \n",
      "***********\n",
      "Position: \n",
      "***********\n",
      "Position:  1\n",
      " 0\n",
      " 3\n",
      " 2\n",
      "\n",
      "Total time of position \n",
      "Total time of position \n",
      "Total time of position \n",
      "Total time of position  1  is:  0.028577089309692383\n",
      " 0  is:  0.02893686294555664\n",
      " 3  is:  0.027788162231445312\n",
      " 2  is:  0.034101009368896484\n",
      "\n",
      "******\n",
      "Analysing Real DBC\n",
      "Pos idx of process is:  4\n",
      "Process id:  95768 . Uname:  posix.uname_result(sysname='Darwin', nodename='Hazals-MBP.lan', release='19.2.0', version='Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64', machine='x86_64')\n",
      "\n",
      "***********\n",
      "Position:  4\n",
      "\n",
      "Total time of position  4  is:  0.00704503059387207\n",
      "\n",
      "*****\n",
      "Calculating inferred distances...\n",
      "\n",
      "*****\n",
      "Calculating inferred distances finished...\n",
      "\n",
      "******\n",
      "Analysing DBC for parameters: \n",
      "******\n",
      "Analysing DBC for parameters: \n",
      "******\n",
      "Analysing DBC for parameters: \n",
      "******\n",
      "Analysing DBC for parameters:  2e-05 0.3 1.0 1.0\n",
      " 2e-05 0.3 1.0 1.0\n",
      " 2e-05 0.3 1.0 1.0\n",
      " 2e-05 0.3 1.0 1.0\n",
      "Pos idx of process is:  0\n",
      "Pos idx of process is:  1\n",
      "Pos idx of process is:  3\n",
      "Pos idx of process is:  2\n",
      "Process id:  95772 . Uname:  posix.uname_result(sysname='Darwin', nodename='Hazals-MBP.lan', release='19.2.0', version='Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64', machine='x86_64')\n",
      "Process id:  95773 . Uname:  posix.uname_result(sysname='Darwin', nodename='Hazals-MBP.lan', release='19.2.0', version='Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64', machine='x86_64')\n",
      "Process id:  95775 . Uname:  posix.uname_result(sysname='Darwin', nodename='Hazals-MBP.lan', release='19.2.0', version='Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64', machine='x86_64')\n",
      "Process id:  95774 . Uname:  posix.uname_result(sysname='Darwin', nodename='Hazals-MBP.lan', release='19.2.0', version='Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64', machine='x86_64')\n",
      "\n",
      "***********\n",
      "Position: \n",
      "***********\n",
      "Position: \n",
      "***********\n",
      "Position: \n",
      "***********\n",
      "Position:  0\n",
      " 1\n",
      " 3\n",
      " 2\n",
      "\n",
      "Loading read position dictionary...\n",
      "Loading read position dictionary...\n",
      "Loading read position dictionary...\n",
      "Loading read position dictionary...\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Loading log_ZCY position dictionary...\n",
      "Loading log_ZCY position dictionary...\n",
      "Loading log_ZCY position dictionary...\n",
      "Loading log_ZCY position dictionary...\n",
      "\n",
      "\n",
      "\n",
      "\tComputing for z_ind:  0\n",
      "\tComputing for z_ind:  0\n",
      "\tComputing for z_ind:  0\n",
      "\tComputing for z_ind:  0\n",
      "\tComputing for z_ind:  1\n",
      "\tComputing for z_ind:  1\n",
      "\tComputing for z_ind:  1\n",
      "\tComputing for z_ind:  1\n",
      "\tComputing for z_ind:  2\n",
      "\tComputing for z_ind:  2\n",
      "\tComputing for z_ind:  2\n",
      "\tComputing for z_ind:  2\n",
      "\tComputing for z_ind:  3\n",
      "\tComputing for z_ind:  3\n",
      "\tComputing for z_ind:  3\n",
      "\tComputing for z_ind:  3\n",
      "\tComputing for z_ind:  4\n",
      "\tComputing for z_ind:  4\n",
      "\tComputing for z_ind:  4\n",
      "\tComputing for z_ind:  4\n",
      "\tComputing for z_ind:  5\n",
      "\tComputing for z_ind:  5\n",
      "\tComputing for z_ind:  5\n",
      "\tComputing for z_ind:  5\n",
      "\tComputing for z_ind:  6\n",
      "\tComputing for z_ind:  6\n",
      "\tComputing for z_ind:  6\n",
      "\tComputing for z_ind:  6\n",
      "\tComputing for z_ind:  7\n",
      "\tComputing for z_ind:  7\n",
      "\tComputing for z_ind:  7\n",
      "\tComputing for z_ind:  7\n",
      "\tComputing for z_ind:  8\n",
      "\tComputing for z_ind:  8\n",
      "\tComputing for z_ind:  8\n",
      "\tComputing for z_ind:  8\n",
      "\tComputing for z_ind:  9\n",
      "\tComputing for z_ind:  9\n",
      "\tComputing for z_ind:  9\n",
      "\tComputing for z_ind:  9\n",
      "\tComputing for z_ind:  10\n",
      "\tComputing for z_ind:  10\n",
      "\tComputing for z_ind:  10\n",
      "\tComputing for z_ind:  10\n",
      "\tComputing for z_ind:  11\n",
      "\tComputing for z_ind:  11\n",
      "\tComputing for z_ind:  11\n",
      "\tComputing for z_ind:  11\n",
      "Total time:  0.01813817024230957\n",
      "Total time:  0.016216039657592773\n",
      "Total time:  0.021708965301513672\n",
      "Total time:  0.01826310157775879\n",
      "\n",
      "***Common Z results:\n",
      "\n",
      "***Common Z results:\n",
      "\n",
      "***Common Z results:\n",
      "\n",
      "***Common Z results:\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Normalized probabilities of common mutation type Z: \n",
      "\n",
      "Normalized probabilities of common mutation type Z: \n",
      "\n",
      "Normalized probabilities of common mutation type Z: \n",
      "\n",
      "Normalized probabilities of common mutation type Z: \n",
      " [6.58956756e-11 6.58956756e-11 1.02494378e-10 6.58932812e-11\n",
      " 6.65364402e-11 6.58932812e-11 1.02563158e-10 6.59044226e-11\n",
      " 6.59044226e-11 1.09836525e-10 9.99999999e-01 1.09836525e-10] [1.03761082e-10 6.67073060e-11 6.67073060e-11 1.11172288e-10\n",
      " 9.99999999e-01 1.11172288e-10 6.67072489e-11 6.67072489e-11\n",
      " 7.78314701e-11 6.67072487e-11 7.78304595e-11 6.67072487e-11] [1.03815687e-10 6.67318384e-11 6.67318384e-11 6.67284908e-11\n",
      " 6.67225415e-11 6.67225415e-11 6.67217046e-11 6.67217046e-11\n",
      " 1.03781962e-10 9.99999999e-01 1.11188200e-10 1.11188200e-10] [2.01384837e-07 2.01385086e-07 2.01384837e-07 9.99997328e-01\n",
      " 2.41686869e-07 2.41686869e-07 2.41679476e-07 3.76023096e-07\n",
      " 2.41679476e-07 2.41690558e-07 2.41676389e-07 2.41676389e-07]\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Maximum probability: \n",
      "Maximum probability: \n",
      "Maximum probability: \n",
      "Maximum probability:  0.9999999991133461 . Dict key:  10\n",
      " 0.9999999991179886 . Dict key:  4\n",
      " 0.9999999991029453 . Dict key:  9\n",
      " 0.9999973280461176 . Dict key:  3\n",
      "\n",
      "Corresponding mutation type: \n",
      "Corresponding mutation type: \n",
      "Corresponding mutation type: \n",
      "Corresponding mutation type:  [2 2] [0 0]\n",
      " [2 1] [3 3]\n",
      " [1 1] [0 2]\n",
      " [0 1] [3 0]\n",
      "\n",
      "Log probabilities of common mutation type Z: \n",
      "\n",
      "Log probabilities of common mutation type Z: \n",
      "\n",
      "Log probabilities of common mutation type Z: \n",
      "\n",
      "Log probabilities of common mutation type Z: \n",
      " [-49.71028796505811, -49.71028796505811, -49.268552831155255, -49.71032430127595, -49.700611012692875, -49.71032430127595, -49.26788199689889, -49.710155234181755, -49.710155234181755, -49.19936766339959, -26.267339668551095, -49.19936766339959]\n",
      " [-40.7579164436034, -41.19969293021904, -41.19969293021904, -40.68892626801228, -17.768986297624156, -40.68892626801228, -41.199693786409796, -41.199693786409796, -41.04546156364864, -41.19969378975043, -41.0454745473923, -41.19969378975043]\n",
      " [-40.641353949289794, -41.08328886128611, -41.08328886128611, -41.08333902793868, -41.083428189609634, -41.083428189609634, -41.08344073211493, -41.08344073211493, -40.641678854568475, -17.652949923361206, -40.57274677765114, -40.57274677765114]\n",
      " [-40.09395322110208, -40.09395198300718, -40.09395322110208, -24.675907744860602, -39.91152794961014, -39.91152794961014, -39.91155854193448, -39.469520343174096, -39.91155854193448, -39.911512688414085, -39.911571312536616, -39.911571312536616]\n",
      "\n",
      "sum Z: \n",
      "\n",
      "sum Z: \n",
      "\n",
      "sum Z: \n",
      "\n",
      "sum Z: \n",
      " 3.91056337096597e-12\n",
      " 1.9187893632853462e-08\n",
      " 2.1548709149108153e-08\n",
      " 1.9203951075120483e-11\n",
      "\n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos:  0  are ids:  1 0  dist:  0.9999900651308726\n",
      " 1  are ids:  1 0  dist:  0.5000480287513362\n",
      " 3  are ids:  1 0  dist:  0.9999899960860544\n",
      " 2  are ids:  2 0  dist:  0.9999868416345264\n",
      "\n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos:  0  are ids:  2 0  dist:  0.5020832041038042\n",
      " 1  are ids:  2 0  dist:  0.5000556766516344\n",
      " 3  are ids:  2 0  dist:  3.5716228557942674e-05\n",
      " 2  are ids:  3 0  dist:  1.56325021262981e-05\n",
      "\n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos:  0  are ids:  2 1  dist:  0.5020832032738066\n",
      " 1  are ids:  2 1  dist:  0.9999486766275917\n",
      " 3  are ids:  2 1  dist:  3.571478144568052e-05\n",
      " 2  are ids:  3 2  dist:  1.5536028962373164e-05\n",
      "\n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos:  0  are ids:  3 0  dist:  0.0009021609040872777\n",
      " 1  are ids:  3 0  dist:  0.499958368107348\n",
      " 3  are ids:  3 0  dist:  0.4999548042264868\n",
      " 2  are ids:  4 0  dist:  0.012088016289448793\n",
      "\n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos:  0  are ids:  3 1  dist:  0.0009021584024389854\n",
      " 1  are ids:  3 1  dist:  1.9190647021358516e-05\n",
      " 3  are ids:  3 1  dist:  0.4999548037440338\n",
      " 2  are ids:  4 2  dist:  0.012087925361354288\n",
      "\n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos:  0  are ids:  3 2  dist:  0.49822672358735215\n",
      " 1  are ids:  3 2  dist:  4.214049104751663e-05\n",
      " 3  are ids:  3 2  dist:  0.5000570994254698\n",
      " 2  are ids:  4 3  dist:  0.9879143588530082\n",
      "\n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: \n",
      "Total time of position  0  are ids:  4 0  dist:  0.01190856976292378\n",
      " 1  are ids:  4 0  dist:  0.49995836780541575\n",
      " 3  are ids:  4 0  dist:  1.5006060871914709e-05\n",
      " 2  is:  0.160477876663208\n",
      "\n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos:  0  are ids:  4 1  dist:  0.01190856728594507\n",
      " 1  are ids:  4 1  dist:  1.9189919299852295e-05\n",
      " 3  are ids:  4 1  dist:  1.5004613645096167e-05\n",
      "\n",
      "******\n",
      "Analysing DBC for parameters: \n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos:  2e-05 0.3 1.0 1.0\n",
      " 0  are ids:  4 2  dist:  0.5018945415488348\n",
      " 1  are ids:  4 2  dist:  4.2139763337692586e-05\n",
      " 3  are ids:  4 2  dist:  0.9999592860709272\n",
      "Pos idx of process is:  4\n",
      "\n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: \n",
      " Similarity. Cells with reads for pos: Process id:  95774 . Uname:  posix.uname_result(sysname='Darwin', nodename='Hazals-MBP.lan', release='19.2.0', version='Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64', machine='x86_64')\n",
      " 0  are ids:  4 3  dist:  0.9872549237729928\n",
      " 1  are ids:  4 3  dist:  0.9999899970855118\n",
      " 3  are ids:  4 3  dist:  0.5000501978809833\n",
      "\n",
      "***********\n",
      "Position: \n",
      "Total time of position \n",
      "Total time of position \n",
      "Total time of position  4\n",
      " 0  is:  0.20948314666748047\n",
      " 1  is:  0.21871709823608398\n",
      " 3  is:  0.23408102989196777\n",
      "\n",
      "Loading read position dictionary...\n",
      "\n",
      "Loading log_ZCY position dictionary...\n",
      "\tComputing for z_ind:  0\n",
      "\tComputing for z_ind:  1\n",
      "\tComputing for z_ind:  2\n",
      "\tComputing for z_ind:  3\n",
      "\tComputing for z_ind:  4\n",
      "\tComputing for z_ind:  5\n",
      "\tComputing for z_ind:  6\n",
      "\tComputing for z_ind:  7\n",
      "\tComputing for z_ind:  8\n",
      "\tComputing for z_ind:  9\n",
      "\tComputing for z_ind:  10\n",
      "\n",
      "*****\n",
      "Plotting distance matrices...\n",
      "\tComputing for z_ind:  11\n",
      "Total time:  0.018449068069458008\n",
      "\n",
      "***Common Z results:\n",
      "\n",
      "\n",
      "Normalized probabilities of common mutation type Z: \n",
      " [1.40668880e-09 1.43050509e-09 2.34444226e-09 1.40675937e-09\n",
      " 1.40665145e-09 1.40665145e-09 1.12562855e-09 1.12537408e-09\n",
      " 1.13039935e-09 9.99999983e-01 1.87493978e-09 1.87493978e-09]\n",
      "\n",
      "Maximum probability:  0.9999999834670201 . Dict key:  9\n",
      "\n",
      "Corresponding mutation type:  [3 1] [0 2]\n",
      "\n",
      "Log probabilities of common mutation type Z: \n",
      " [-47.72964866503542, -47.71285964299041, -47.2188397032845, -47.72959849649559, -47.72967521697812, -47.72967521697812, -47.95254564644442, -47.95277173982675, -47.94831625704173, -27.347621415931275, -47.442310692592926, -47.442310692592926]\n",
      "\n",
      "sum Z: \n",
      " 1.3276357337198352e-12\n",
      "\n",
      " Similarity. Cells with reads for pos:  4  are ids:  1 0  dist:  0.5000149525988195\n",
      "\n",
      " Similarity. Cells with reads for pos:  4  are ids:  2 0  dist:  0.9999803172253624\n",
      "\n",
      " Similarity. Cells with reads for pos:  4  are ids:  2 1  dist:  0.5000148454501201\n",
      "\n",
      " Similarity. Cells with reads for pos:  4  are ids:  3 0  dist:  1.5017799936581305e-05\n",
      "\n",
      " Similarity. Cells with reads for pos:  4  are ids:  3 1  dist:  0.4999900577893195\n",
      "\n",
      " Similarity. Cells with reads for pos:  4  are ids:  3 2  dist:  1.469634193427224e-05\n",
      "\n",
      " Similarity. Cells with reads for pos:  4  are ids:  4 0  dist:  9.438056762338787e-05\n",
      "\n",
      " Similarity. Cells with reads for pos:  4  are ids:  4 1  dist:  0.5000165120491848\n",
      "\n",
      " Similarity. Cells with reads for pos:  4  are ids:  4 2  dist:  9.405913342898738e-05\n",
      "\n",
      " Similarity. Cells with reads for pos:  4  are ids:  4 3  dist:  0.9999106325940454\n",
      "\n",
      "Total time of position  4  is:  0.1192779541015625\n"
     ]
    }
   ],
   "source": [
    "%run analyse_dbc.py ../data/2020_07_09/  --seed_val 42 --p_ae 2e-05 --pos_range_max 0 --b_g 1 --print_status 0 --data_type synthetic --p_ado 0.3 --output_dir ../results/2020_07_09_results/ --pos_range_min 0 --a_g 1 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combine distance between cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "*****\n",
      "Loading dictionaries\n",
      "\n",
      "Total time of loading dictionaries:  0.0044481754302978516\n",
      "\n",
      "*****\n",
      "Combining multiple sites\n",
      "\n",
      "*****\n",
      "Combining multiple sites\n",
      "\n",
      "Using  1  positions\n",
      "\n",
      "Using  2  positions\n",
      "\n",
      "Using  3  positions\n",
      "\n",
      "Using  4  positions\n",
      "\n",
      "Using  5  positions\n",
      "\n",
      "*****\n",
      "Analysis: \n",
      "\tTotal number: \t 5\n",
      "\tHeterozygous: \t 5 . \t% 100.0\n",
      "\tCorrect heterozygous: \t 0 . \t% 0.0\n",
      "\tHomozygous: \t 0 . \t% 0.0\n",
      "\t(Skipped positions:  [] )\n",
      "\n",
      "Total time of combining multiple sites:  5.861721038818359\n",
      "\n",
      "Total time of combining sites:  5.862156867980957\n"
     ]
    }
   ],
   "source": [
    "%run analyse_dbc_combine.py ../data/2020_07_09/  --pos_range_max 0 --data_type synthetic --output_dir ../results/2020_07_09_results/ --pos_range_min 0 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reconstruct lineage tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Trees of case:  selected \n",
      "\n",
      "\n",
      "\tCase:  selected .\tTree with bulk as a node:\n",
      "\n",
      "/--------------------------------------------------------------------------- V2\n",
      "|                                                                              \n",
      "+                                                        /------------------ V0\n",
      "|                  /-------------------------------------+                     \n",
      "|                  |                                     \\------------------ V1\n",
      "\\------------------+                                                           \n",
      "                   |                  /------------------------------------- V5\n",
      "                   \\------------------+                                        \n",
      "                                      |                  /------------------ V3\n",
      "                                      \\------------------+                     \n",
      "                                                         \\------------------ V4\n",
      "                                                                               \n",
      "                                                                               \n",
      "[&R] (V2:1.3877787807814457e-17,((V0:0.0,V1:0.0):0.40000000000000013,(V5:-5.551115123125783e-17,(V3:0.0,V4:0.0):0.6000000000000001):-1.6653345369377348e-16):1.3877787807814457e-17);\n",
      "\n",
      "\n",
      "\tCase:  selected .\tTree with bulk as a root:\n",
      "\n",
      "                                                  /------------------------- V3\n",
      "/-------------------------------------------------+                            \n",
      "|                                                 \\------------------------- V4\n",
      "+                                                                              \n",
      "|                                                 /------------------------- V0\n",
      "|                        /------------------------+                            \n",
      "\\------------------------+                        \\------------------------- V1\n",
      "                         |                                                     \n",
      "                         \\-------------------------------------------------- V2\n",
      "                                                                               \n",
      "                                                                               \n",
      "[&R] ((V3:0.0,V4:0.0):0.6000000000000001,((V0:0.0,V1:0.0):0.40000000000000013,V2:2.7755575615628914e-17):-1.6653345369377348e-16)V5;\n",
      "\n",
      "\n",
      "\tCase:  selected .\tTree with bulk, midpoint rooted:\n",
      "\n",
      "                                                  /------------------------- V3\n",
      "/-------------------------------------------------+                            \n",
      "|                                                 \\------------------------- V4\n",
      "+                                                                              \n",
      "|                                                 /------------------------- V0\n",
      "|                        /------------------------+                            \n",
      "\\------------------------+                        \\------------------------- V1\n",
      "                         |                                                     \n",
      "                         \\-------------------------------------------------- V2\n",
      "                                                                               \n",
      "                                                                               \n",
      "[&R] ((V3:0.0,V4:0.0):0.5,((V0:0.0,V1:0.0):0.40000000000000013,V2:2.7755575615628914e-17):0.09999999999999992);\n",
      "\n",
      "\n",
      "\tCase:  selected .\tTree without bulk:\n",
      "\n",
      "                                                  /------------------------- V0\n",
      "/-------------------------------------------------+                            \n",
      "|                                                 \\------------------------- V1\n",
      "+                                                                              \n",
      "|                        /-------------------------------------------------- V2\n",
      "\\------------------------+                                                     \n",
      "                         |                        /------------------------- V4\n",
      "                         \\------------------------+                            \n",
      "                                                  \\------------------------- V3\n",
      "                                                                               \n",
      "                                                                               \n",
      "[&R] ((V0:0.0,V1:0.0):0.19999999999999998,(V2:-1.1102230246251565e-16,(V4:0.0,V3:0.0):0.6000000000000001):0.19999999999999998);\n",
      "\n",
      "\n",
      "\tCase:  selected .\tTree without bulk, midpoint rooted:\n",
      "\n",
      "                                                  /------------------------- V4\n",
      "/-------------------------------------------------+                            \n",
      "|                                                 \\------------------------- V3\n",
      "+                                                                              \n",
      "|                        /-------------------------------------------------- V2\n",
      "\\------------------------+                                                     \n",
      "                         |                        /------------------------- V0\n",
      "                         \\------------------------+                            \n",
      "                                                  \\------------------------- V1\n",
      "                                                                               \n",
      "                                                                               \n",
      "[&R] ((V4:0.0,V3:0.0):0.5,(V2:-1.1102230246251565e-16,(V0:0.0,V1:0.0):0.39999999999999997):0.10000000000000009);\n",
      "\n",
      "\n",
      "Trees of case:  inferred \n",
      "\n",
      "\n",
      "\tCase:  inferred .\tTree with bulk as a node:\n",
      "\n",
      "                                                  /------------------------- V2\n",
      "/-------------------------------------------------+                            \n",
      "|                                                 \\------------------------- V5\n",
      "|                                                                              \n",
      "+                                                 /------------------------- V3\n",
      "|                        /------------------------+                            \n",
      "|                        |                        \\------------------------- V4\n",
      "\\------------------------+                                                     \n",
      "                         |                        /------------------------- V0\n",
      "                         \\------------------------+                            \n",
      "                                                  \\------------------------- V1\n",
      "                                                                               \n",
      "                                                                               \n",
      "[&R] ((V2:0.04664885417499988,V5:0.053784193925000126):0.06894537630625003,((V3:0.050710362600000045,V4:0.05426561539999995):0.4588116947624999,(V0:0.09361112120833343,V1:0.15637811819166658):0.19290858756250012):0.06894537630625003);\n",
      "\n",
      "\n",
      "\tCase:  inferred .\tTree with bulk as a root:\n",
      "\n",
      "/--------------------------------------------------------------------------- V2\n",
      "|                                                                              \n",
      "+                                                 /------------------------- V3\n",
      "|                        /------------------------+                            \n",
      "|                        |                        \\------------------------- V4\n",
      "\\------------------------+                                                     \n",
      "                         |                        /------------------------- V0\n",
      "                         \\------------------------+                            \n",
      "                                                  \\------------------------- V1\n",
      "                                                                               \n",
      "                                                                               \n",
      "[&R] (V2:0.04664885417499988,((V3:0.050710362600000045,V4:0.05426561539999995):0.4588116947624999,(V0:0.09361112120833343,V1:0.15637811819166658):0.19290858756250012):0.13789075261250006)V5;\n",
      "\n",
      "\n",
      "\tCase:  inferred .\tTree with bulk, midpoint rooted:\n",
      "\n",
      "                                                  /------------------------- V3\n",
      "/-------------------------------------------------+                            \n",
      "|                                                 \\------------------------- V4\n",
      "+                                                                              \n",
      "|                                                 /------------------------- V0\n",
      "|                        /------------------------+                            \n",
      "\\------------------------+                        \\------------------------- V1\n",
      "                         |                                                     \n",
      "                         \\-------------------------------------------------- V2\n",
      "                                                                               \n",
      "                                                                               \n",
      "[&R] ((V3:0.050710362600000045,V4:0.05426561539999995):0.3769163925583333,((V0:0.09361112120833343,V1:0.15637811819166658):0.19290858756250012,V2:0.18453960678749992):0.08189530220416658);\n",
      "\n",
      "\n",
      "\tCase:  inferred .\tTree without bulk:\n",
      "\n",
      "                                                  /------------------------- V4\n",
      "/-------------------------------------------------+                            \n",
      "|                                                 \\------------------------- V3\n",
      "+                                                                              \n",
      "|                        /-------------------------------------------------- V2\n",
      "\\------------------------+                                                     \n",
      "                         |                        /------------------------- V0\n",
      "                         \\------------------------+                            \n",
      "                                                  \\------------------------- V1\n",
      "                                                                               \n",
      "                                                                               \n",
      "[&R] ((V4:0.07156110758333327,V3:0.03341487041666673):0.25586449470625,(V2:0.18453960678749987,(V0:0.10917072268750003,V1:0.14081851671249998):0.13999129291250018):0.25586449470625);\n",
      "\n",
      "\n",
      "\tCase:  inferred .\tTree without bulk, midpoint rooted:\n",
      "\n",
      "                         /-------------------------------------------------- V2\n",
      "/------------------------+                                                     \n",
      "|                        |                        /------------------------- V0\n",
      "|                        \\------------------------+                            \n",
      "+                                                 \\------------------------- V1\n",
      "|                                                                              \n",
      "|                                                 /------------------------- V4\n",
      "\\-------------------------------------------------+                            \n",
      "                                                  \\------------------------- V3\n",
      "                                                                               \n",
      "                                                                               \n",
      "[&R] ((V2:0.18453960678749987,(V0:0.10917072268750003,V1:0.14081851671249998):0.13999129291250018):0.15124014368541655,(V4:0.07156110758333327,V3:0.03341487041666673):0.36048884572708345);\n",
      "\n",
      "\n",
      "\n",
      "Comparison Results for inferred and selected for _w_bulk_root.tre\n",
      "R-F:  2\n",
      "Normalised R-F:  0.125\n",
      "Weighted R-F:  0.8877845418625003\n",
      "Normalised Weighted R-F:  0.05548653386640627\n",
      "Euc:  0.350348999778636\n",
      "\n",
      "\n",
      "Comparison Results for inferred and selected for _w_bulk_root_midpoint.tre\n",
      "R-F:  0\n",
      "Normalised R-F:  0.0\n",
      "Weighted R-F:  0.8877845418624999\n",
      "Normalised Weighted R-F:  0.05548653386640624\n",
      "Euc:  0.36214987206209104\n",
      "\n",
      "\n",
      "Comparison Results for inferred and selected for _wo_bulk.tre\n",
      "R-F:  2\n",
      "Normalised R-F:  0.125\n",
      "Weighted R-F:  1.399513531275\n",
      "Normalised Weighted R-F:  0.0874695957046875\n",
      "Euc:  0.5473001407631862\n",
      "\n",
      "\n",
      "Comparison Results for inferred and selected for _wo_bulk_midpoint.tre\n",
      "R-F:  0\n",
      "Normalised R-F:  0.0\n",
      "Weighted R-F:  0.9902648292333328\n",
      "Normalised Weighted R-F:  0.0618915518270833\n",
      "Euc:  0.40216160402355383\n"
     ]
    }
   ],
   "source": [
    "%run lineage_trees.py ../results/2020_07_09_results/ --data_type synthetic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
